### DISCUSSION TITLE
Facial recognition example? Contact info? and you now have a dependent on NPM

### DISCUSSION THREAD

**Discussion by meeki007** [vladmandic/face-api#23] [Show and tell]
<b>Facial recognition example:</b>

I've tried many times now to get a vanilla nodeJS environment to work with your example flow but its not been a priority as I've been busy with my own repo. However I do have a working version of facial recognition that works with your face-api in node-red.

If you take a look at my [facial-recognition.js](https://github.com/meeki007/node-red-contrib-facial-recognition/blob/main/facial-recognition.js) line in code 771 to 828.  It allows a user to just name the folder and put images into it. Thought you might find this helpful.

<b>Your Contact info </b>

I wanted to email you a link to a server running the node-red version of Facial recognition based on your face-api so you could play around with it if you wanted, also so i did not have to blast/treat your issues section as a message board.
I think I found your Instagram do you want me to send you a message there?

<b>you now have a dependent on NPM</b>
[https://www.npmjs.com/browse/depended/@vladmandic/face-api](https://www.npmjs.com/browse/depended/@vladmandic/face-api)

I'm proud to be the first one on the list

Thanks again,

Meeki

I'm just so tickled i got this working:

![image.jpg](https://raw.githubusercontent.com/meeki007/node-red-contrib-facial-recognition/main/doc/image.jpg)

**Comment by vladmandic** [comment]
> I wanted to email you a link to a server running the node-red version of Facial recognition based on your face-api
> so you could play around with it if you wanted, also so i did not have to blast/treat your issues section as a message board.

I'll definitely check it out! Never played with `node-red` myself.

> I think I found your Instagram do you want me to send you a message there?

Yes, you have :)

On the topic of non-issue messages, **github** recently allows for **discussions** & i've just enabled it (see top menu line, close to issues)!

(and i've moved this issue to be a first post in discussions)

**Comment by vladmandic** [comment]
> If you take a look at my facial-recognition.js line in code 771 to 828. It allows a user to just name the folder and put images into it. 

Yes, I use similar approach in my apps, just didn't see the need to put it in demo app as it uses fixed demo images.

**Reply by meeki007** [reply]
Did you know there are helpers for de/-serializing LabeledFaceDescriptors
https://github.com/justadudewhohacks/face-api.js/pull/397

I'm wondering if I should add that to your documentation. That little gem saved me some serious time.
Or if any one can code to that level they will find it like it did ? 

Or just wait for someone to ask about persisting the LabeledFaceDescriptors()

**Comment by vladmandic** [comment]
> Did you know there are helpers for de/-serializing LabeledFaceDescriptors

I did see that, but didn't find it a complete solution.  
I implemented my serializer in the app that uses `face-api` library so I can store **all** results in a database.  

In hindsight, it should probably be moved from app level to library level.  
(like you said, it would save some serious time)  

```js
function flattenObject(object) {
  const stripped = {};
  for (const key of Object.keys(object)) {
    if (key[0] === '_') stripped[key.substr(1)] = object[key];
    else stripped[key] = object[key];
  }
  return stripped;
}

async function faceapiClassify(image) {
  const results = await models.faceapi.detectAllFaces(image, models.faceapi.options)
    .withFaceLandmarks()
    .withFaceExpressions()
    .withFaceDescriptors()
    .withAgeAndGender();
  const faces = [];
  for (const result of results) {
    const emotion = Object.entries(result.expressions).reduce(([keyPrev, valPrev], [keyCur, valCur]) => (valPrev > valCur ? [keyPrev, valPrev] : [keyCur, valCur]));
    const object = {
      scoreGender: result.genderProbability,
      gender: result.gender,
      age: result.age,
      descriptor: result.descriptor,
      scoreEmotion: emotion && emotion[1] ? emotion[1] : 0,
      emotion: emotion && emotion[0] ? emotion[0] : '',
      box: flattenObject(result.detection.box),
      points: result.landmarks.positions.map((a) => flattenObject(a)),
    };
    faces.push(object);
  }
  return faces;
}
```

**Reply by meeki007** [reply]
without seeing more of your code and how your calling the function `faceapiClassify` I don't know how you would use it.

Something like
```JS
async function faceapiClassify(image) {
  const results = await models.faceapi.detectAllFaces(image, models.faceapi.options)
    .withFaceLandmarks()
    .withFaceExpressions()
    .withFaceDescriptors()
    .withAgeAndGender();

    let results_toJson = results.map(x=>x.toJSON());
  }
  return results_toJson;
}
```

Here is how im using it. (i pulled out a bunch of checks from the function to shrink it. this is a parsed down version)
```JS
function LoadLabeledImages()
{

  return Promise.all(
    list_dirs_in_labeled_face_folder_names_only.map(async each_dir_name => {
      const image_names_in_each_dir_name = await listFiles(path.join(user_path, each_dir_name)).catch(error => {
        notify_user_errors(error);
      });
      const length = image_names_in_each_dir_name.length
      const descriptions = [];
      for (let i = 0; i < length; i++)
      {
        const image_file_size = fs.statSync(image_names_in_each_dir_name[i]);
        const bufferd_img = await fs.readFileSync(image_names_in_each_dir_name[i]);
        const face_detect_tensor = await image(bufferd_img);
        var detections;
          if ( face_detect_tensor )
          {
            detections = await faceapi.detectSingleFace(face_detect_tensor).withFaceLandmarks().withFaceDescriptor();
            descriptions.push(detections.descriptor);
            face_detect_tensor.dispose();
          }
        }
      }
      return new faceapi.LabeledFaceDescriptors(each_dir_name, descriptions);
    })
  );
}

const labeledFaceDescriptors = await LoadLabeledImages();

//use helper toJSON so users can save the FaceDescriptors
let labeledFaceDescriptors_toJson = labeledFaceDescriptors.map(x=>x.toJSON());
nodeContext.set('labeledFaceDescriptors',labeledFaceDescriptors_toJson);
```

and if I want to load it back in 
```JS
context_labeledFaceDescriptors = nodeContext.get('labeledFaceDescriptors');
let labeledFaceDescriptors_fromJson = context_labeledFaceDescriptors.map( x=>faceapi.LabeledFaceDescriptors.fromJSON(x));
const faceMatcher = await new faceapi.FaceMatcher(labeledFaceDescriptors_fromJson, this.Face_Recognition_distanceThreshold);
```

**Reply by meeki007** [reply]
So yep; in the simplest form of use.

> For community:
> `var labeledFaceDescriptorsJson = labeledFaceDescriptors.map(x=>x.toJSON())`
> and (not tested, but i think it works actually like this one)
> `var labeledFaceDescriptors = labeledFaceDescriptorsJson.map( x=>faceapi.LabeledFaceDescriptors.fromJSON(x) );`

https://github.com/justadudewhohacks/face-api.js/pull/397#issuecomment-526847142

**Comment by vladmandic** [comment]
that works.

my function remaps everything (not just descriptors) to structure i like and then if i want to store it, it's simple `JSON.stringify()`.
remapping everything with `.toJSON()` works, but resulting object has a lot of unnecessary items.  
it's just a question of preference.
